metagenomics.py

```diff
@@ -1143,6 +1144,162 @@ class KrakenBuildError(Exception):
+                        choices=["species", "genus", "family", "order", "class", "phylum", "kingdom", "superkingdom"])
+    parser.add_argument('--topN', type=int, dest="top_n_entries", help='Only include the top N taxa by read count (default: %(default)s)')
+    parser.add_argument('--countThreshold', type=int, dest="count_threshold", help='Minimum number of reads to be included (default: %(default)s)', default=1)
+    parser.add_argument('--noHist', action='store_true', dest="no_hist", help='When topN==1, write out a report by-sample rather than a histogram.')
+    util.cmd.common_args(parser, (('loglevel', None), ('version', None), ('tmp_dir', None)))
+    util.cmd.attach_main(parser, taxlevel_summary, split_args=True)
+    return parser
+
+def taxlevel_summary(summary_files_in, json_out, csv_out, tax_headings, taxlevel_focus, top_n_entries, count_threshold, no_hist):
+    """
+        By default, when --taxHeading is at the same level as --taxlevelFocus
+        a summary with lines for each sample is emitted. Otherwise, a histogram is returned.
+        Full data returned via --jsonOut (filtered by --topN and --countThreshold), whereas -csvOut returns a summary.
+    """
+
+    samples = {}
+    same_level = False
+    
+    def indent_len(in_string):
+        return len(in_string)-len(in_string.lstrip())
+
+    for f in list(summary_files_in):
+        sample_name, extension = os.path.splitext(f)
+        sample_summary = {}
+        tax_headings_copy = [s.lower() for s in tax_headings]
+
+        with util.file.open_or_gzopen(f, 'rU') as inf:
+            should_process = False
+            indent_of_selection = -1
+            currently_being_processed = ""
+            for line in inf:
+                if len(line.rstrip('\r\n').strip()) == 0:
+                    continue
+                csv.register_dialect('kraken_report', quoting=csv.QUOTE_MINIMAL, delimiter="\t")
+                row = next(csv.reader([line.strip().rstrip('\n')], dialect="kraken_report"))
+                
+                indent_of_line = indent_len(row[5])
+                row = [item.strip() for item in row] # remove leading/trailing whitespace from each item
+                
+                # rows are formatted like so:
+                # 0.00  16  0   D   10239     Viruses
+                #
+                # row[0] Percentage of reads covered by the clade rooted at this taxon
+                # row[1] Number of reads covered by the clade rooted at this taxon
+                # row[2] Number of reads assigned directly to this taxon
+                # row[3] A rank code, indicating (U)nclassified, (D)omain, (K)ingdom, (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. All other ranks are simply '-'.
+                # row[4] NCBI taxonomy ID
+                # row[5] indented scientific name
+
+                if indent_of_line == indent_of_selection:
```

Comment:
After finishing all the children selection, there's no guarantee the next line will have the same indent as the selection indent, so this should be `<=`

Reply:
@tomkinsc would you like to revive this PR for merging in soon? If so, it looks like of @yesimon 's comments last month, this is the only one about correctness that ought to be address, while the rest are stylistic.

Reply:
Potentially. I delivered some output to Anne for her work, and may incorporate changes based on her suggestions, if any.